### DISCUSSION TITLE
Event racing condition

### DISCUSSION THREAD

**Discussion by krishnak** [micropython/micropython#9544] [Using MicroPython]
I have an asyncio method which sets this
```
responseQueue.append(msg)
                    responseQueueEvent.set()
                    responseQueueEvent.clear()

```                    

The method which is waiting on responseQueueEvent is as below, the responseQueue object is a **global** variable
There are occassions where the event is set but the queue is empty at the waiting method, it is not that the msg variable is empty. The append function appears to be completing after Event.set() is called which results in the waiting function to proceed with an empty list. How do I avoid this

The same code works OK 80% of the time.

 ```
try:
                await asyncio.wait_for(responseQueueEvent.wait(),5)
                print("No time out leaving as response received")
                if len(responseQueue)>0:
                    print("queue has values")

            except asyncio.TimeoutError:  
                print("No response received")
                continue
```

**Comment by karfas** [comment]
What exactly are these `responseQueue` and `responseQueueEvent` objects ?
The asyncio queue found at https://github.com/peterhinch/micropython-async/blob/master/v3/primitives/queue.py has a different interface and it's own event handling.
However, I don't see the the benefit from using an event for the queue. The task who reads from the queue could wait (maybe using `asyncio.sleep(0)`) until at least one entry is in the queue.

**Reply by krishnak** [reply]
responsequeue is a list and the event is an asyncio Event. I have figured out the issue, since the responseQueue has been set as a global variable and is only referenced at two places, i.e once each in the setter async method and in the "waiting" async method, there is a gc.collect call occurring after the variable has been set. This gc event probably removes the object.  I have now moved the global variable inside main which appears to prevent it from being garbage collected.

**Reply by jimmo** [reply]
That doesn't really make sense in terms of how the GC works.

Where are you calling `.set()` from? If it's from an interrupt, you should be using ThreadSafeFlag instead.

**Reply by krishnak** [reply]
pseudo code as below

```
responseQueue = []
responseQueueEvent = asyncio.Event()
async def process_message():
    global responseQueue
    while True:
         try:
                   #wait for message and do stuff based on message
              if message == typeA:
                    responseQueue.append(msg)
                    responseQueueEvent.set()
                    responseQueueEvent.clear()
              elif:
                   some code
              elif:
                    some code
         except Exception as any:
                  some code
          finally:
                 gc.collect()
          await asyncio.sleep(0)

async def anothermethodwhichiswaiting():
     global responseQueue
     try:
                await asyncio.wait_for(responseQueueEvent.wait(),5)
                print("Not timed out, leaving as response received")
                if len(responseQueue)>0:
                    #this should always happen if the event is set but happens 80-90% of the time
                else:
                    print("bad response")   

            except asyncio.TimeoutError: 
                print("No response from remote host WHO HAS timed out")
                
       
```

I just moved the responseQueue object which was not part of any method earlier, to be included inside the main which starts these async methods as a task like this

```
async def main():
global responseQueue
responseQueue = []
await asyncio.create_task(process_message())
await asyncio.create_task(anothermethodwhichiswaiting())
```
With this change the problem appears to be gone.

**Reply by jimmo** [reply]
```
responseQueueEvent.set()
responseQueueEvent.clear()
```

It doesn't make sense to set an event and then immediately clear it. The waiter should clear it.

**Reply by krishnak** [reply]
I too had a similar opinion as you when I saw set/clear,it as done by Peter in his Queue code referred in the previous message. It appears that the set sends a message to all who are all currently in the responsequeue.wait() mode- and a clear immediately is only a way of clearing it in one place rather than by many waiters. But that part works

**Reply by krishnak** [reply]
It makes sense if the waiter has timed out and there is no one to clear a stale event.

**Reply by peterhinch** [reply]
@jimmo My understanding is that it is valid to immediately clear an `Event` if you can guarantee that the waiting task(s) are paused waiting on the `Event` at the time it is set. If the waiting task hasn't reached the `Event` at the time it is cleared, then events will be missed. That said, I too would recommend clearing it in the waiting task.

The symptom we have here is that the `Event` seems to be being set when the queue is empty: too many triggers rather than too few. On the basis of the code presented this doesn't seem to make sense. Nor do the symptoms relating to globals and GC. Something else is going on. Are interrupts or threaded code involved anywhere?

I can see two options to try to debug this:
 1. Use print statements to show queue contents when the event is set and when it is awaited.
 2. Alternatively write a minimal test case and gradually extend it until it becomes the full solution.

**Reply by krishnak** [reply]
Hello Peter

There are no threads or interrupts, just asyncio. When the bug occured - the print statements printing the queue will print expected values - but this will be always printed AFTER the code exits the anothermethodwhichiswaiting - inside this method the queue will be always empty.
However the place where it gets filled, i.e process message - the valid values get filled but it gets printed only after the waiting method exits.

I have tried to set high time outs like 10 seconds - irrespective of that, the queue value always gets printed afterwards. So my theory of GC sounds illogical.

My code has undergone major transformation in the last 24 hours and I have moved away from this problem. I can try to reproduce this problem at a later date and if reproduce-able, I shall update. For time being please treat this as resolved.

**Reply by jimmo** [reply]
@krishnak just a quick note that there seems to be a misunderstanding here about how `global` works.

You only need to use the `global` keyword if your code assigns to a variable. Your code doesn't assign to `responseQueue` -- it calls `.append()` and presuably `.pop()`.

**Reply by krishnak** [reply]
After re-reading your comment, I understood what you are meaning, it appears to be a redundant declaration at the start of each function.

When the issue occured,  I had declared myqueue = [] at the start of the file, I shifted that inside the main method which starts the async tasks. This immediately resolved the issue. To recreate the problem I assigned another queue with my current code similarly at the start of the code when I encountered the problem. I am not able to repoduce the problem.

At that time of my issue, I had a simple code which operated on a single frequency, it has now transformed in to multi frequency hoping code with couple of locks. Hence if at all the issue occurs it should occur now. I appreciate the effort you all show here to resolve the reported issue. I don't want you to waste your time on this unless I am able to reproduce it again.
